keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+}
+
+// ApplicationHealthCheck contains a template for a Pod that performs an Application health check. The readiness 
+// of the Pod decides the readiness of the Application.
+type ApplicationHealthCheck struct {
+    
+    // Template is used to specify a Pod whose readiness will be used to determine the readiness of an Application.
+    Template v1.PodTemplateSpec
+}
+// ApplicationSpec is the specification object for an Application. It specifies the Application's components, 
+// dependencies, and an optional health check. Its selector can be used to select application components.
+type ApplicationSpec struct {
+    // Type is the type of the application (e.g. WordPress, MySQL, Cassandra).
+    Type string `json:"type" protobuf:"bytes,1,name=type"`
+    
+    // Components is a map of the applications components to the kinds created by the application (e.g. Pods, Services, 
+    // Deployments, CRDS, etc)
+    Components map [string]metav1.GroupVersionKind `json:"components,omitempty" protobuf:"bytes,2,opt,name=components"`
+    
+    // Dependencies is a list of other Applications that this Application depends on.
+    Dependencies []sting `json:"dependencies" protobuf:"bytes,3,opt,name=dependencies"`
+    
+    // selector is a label query over kinds that created by the application. It must match the component objects' labels.
+    // More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+    Selector *metav1.LabelSelector `json:"selector" protobuf:"bytes,4,opt,name=selector"`
+    
+    // HealthCheck is an optional application level health check.
+    HealthCheck* ApplicationHealthCheck `json:"healthCheck" protobuf:"bytes,5,opt,name=healthCheck"`
+    
+    // Version is an optional version indicator for the application.
+    Version string `json:"version,omitempty" protobuf:"bytes,6,opt,name=version"`
+    
+    // AboutURL is a URL pointing to any additional information about the application.
+    AboutURL string `json:"url,omitempty" protobuf:"bytes,7,opt,name=url"`
+    
+    // Description is a brief string description of the application.
+    Description string `json:"description,omitempty" protobuf:"bytes,8,opt,name=description"`
+    
+    // Maintainers is an optional list of maintainers of the application
+    Maintainers []string `json:"maintainers,omitempty" protobuf:"bytes,9,opt,name=maintainers"`
+    
+    // Keywords is an optional list of key words associated with the application (e.g. MySQL, RDBMS, database).
+    Keywords []string `json:"keywords,omitempty" protobuf:"bytes,10,opt,name=keywords"`
+    
+    // LicenseURL is a an optional URL pointing to any licensing information for the application.
+    LicenseURL string `json:"license,omitempty" protobuf:"bytes,11,opt,name=license"`
+    
+    // DashboardURL is a an optional URL pointing to the applications dashboard.
+    DashboardURL string `json:"dashboard,omitempty" protobuf:"bytes,12,opt,name=dashboard"`
```

Comment:
Lots of applications have multiple dashboards.  Do I create multiple applications that overlap in order to expose their URLs?

Reply:
Why surface Dashboard and not other URLs that an app may have?

Reply:
I could also imagine links to monitoring and logging dashboards.

It does seem like there may need to be multiple links.

Reply:
Sounds reasonable. My initial thought was that this would be a landing page that aggregates multiple dashboards for the app, but expanding it to be a container of URLs (potentially mapped to the dashboard name) meets both use cases.

Reply:
Also, there may be no exposed Service or Ingress for an Application. What if the Application provides cluster internal functionality and exposes no load balanced Service or Ingress (e.g. RDBMs, MQ, Stream processor).